assets/src/hooks/useVehiclesSocket.ts

```diff
@@ -0,0 +1,175 @@
+  payload: { routeId },
+})
+
+type Action =
+  | SetSocketAction
+  | SetChannelForRouteAction
+  | SetVehiclesForRouteAction
+  | RemoveRouteAction
+
+type Dispatch = ReactDispatch<Action>
+
+export const reducer = (state: State, action: Action): State => {
+  switch (action.type) {
+    case "SET_SOCKET":
+      return {
+        ...state,
+        socket: action.payload.socket,
+      }
+    case "SET_CHANNEL":
+      return {
+        ...state,
+        channelsByRouteId: {
+          ...state.channelsByRouteId,
+          [action.payload.routeId]: action.payload.channel,
+        },
+      }
+    case "SET_VEHICLES_FOR_ROUTE":
+      return {
+        ...state,
+        vehiclesByRouteId: {
+          ...state.vehiclesByRouteId,
+          [action.payload.routeId]: action.payload.vehicles,
+        },
+      }
+    case "REMOVE_ROUTE":
+      const {
+        [action.payload.routeId]: channel,
+        ...channelsWithoutRouteId
+      } = state.channelsByRouteId
+      const {
+        [action.payload.routeId]: vehicles,
+        ...vehiclesWithoutRouteId
+      } = state.vehiclesByRouteId
+      return {
+        ...state,
+        channelsByRouteId: channelsWithoutRouteId,
+        vehiclesByRouteId: vehiclesWithoutRouteId,
+      }
+    default:
+      return state
+  }
+}
+
+const subscribe = (
+  socket: Socket,
+  routeId: RouteId,
+  dispatch: Dispatch
+): Channel => {
+  const handleVehicles = ({ vehicles }: { vehicles: Vehicle[] }) => {
+    dispatch(setVehiclesForRoute(routeId, vehicles))
+  }
+
+  const topic = `vehicles:${routeId}`
+  const channel = socket.channel(topic)
+  channel.on("vehicles", handleVehicles)
+  channel
+    .join()
+    .receive("ok", handleVehicles)
+    .receive("error", ({ reason }) => console.log("join failed", reason))
+    .receive("timeout", () => console.log("join timeout"))
+  return channel
+}
+
+const useVehiclesSocket = (
+  selectedRouteIds: RouteId[]
+): VehiclesByRouteId => {
+  const [state, dispatch] = useReducer(reducer, initialState)
+  const { socket, channelsByRouteId, vehiclesByRouteId } = state
+
+  useEffect(() => {
+    const initialSocket = new Socket("/socket")
+    initialSocket.connect()
+    dispatch(setSocket(initialSocket))
+  }, [])
+
+  useEffect(() => {
+    if (socket) {
+      // Unsubscribe from any routes we don't care about anymore
+      Object.entries(channelsByRouteId).forEach(([routeId, channel]) => {
+        if (!selectedRouteIds.includes(routeId)) {
+          channel.leave()
+          dispatch(removeRoute(routeId))
+        }
+      })
+
+      // Subscribe to any routes we're not already subscribed to
+      selectedRouteIds.forEach((routeId: RouteId) => {
+        if (!(routeId in channelsByRouteId)) {
+          const channel = subscribe(socket, routeId, dispatch)
+          dispatch(setChannelForRoute(routeId, channel))
+        }
+      })
+    }
+  }, [socket, selectedRouteIds])
+
+  return vehiclesByRouteId
+}
+
+export default useVehiclesSocket
```

Comment:
What about splitting this into two hooks? One to maintain the socket, and the other to maintain a channel for each route?

Reply:
Mmm. That was we could reuse the socket elsewhere, and mocking the socket in the tests becomes easier. I like it.